examples/py_triple.rs

```diff
@@ -0,0 +1,411 @@
+// This example creates a single-chip circuit that proves knowledge of two field elements `a` and
+// `b` such that `a^2 + b^2 == PI^2` (for public input `PI`), i.e. proves knowledge of a Pythagorean
+// triple.
+//
+// The arithmetization for this computation is:
+//
+// (0) allocate private inputs: a, b
+// (1) allocate public input: c = PI
+// (2) multiply: a * a = a^2
+// (3) multiply: b * b = b^2
+// (4) multiply: c * c = c^2
+// (5) add: a^2 + b^2 = c^2
+//
+// The constraint system has 3 advice columns `l` (left), `r` (right), and `o` (output), one
+// instance column `pub_col` (contains the public inputs), and 3 selectors (fixed columns) `s_add`
+// (addition gate), `s_mul` (multiplication gate), and `s_pub` (public input gate).
+//
+// |-----|-------|-------|-------|---------|-------|-------|-------|
+// | row | l_col | r_col | o_col | pub_col | s_add | s_mul | s_pub |
+// |-----|-------|-------|-------|---------|-------|-------|-------|
+// |  0  |   a   |   b   |       |   0     |   0   |   0   |   0   |
+// |  1  |   c   |       |       |   PI    |   0   |   0   |   1   |
+// |  2  |   a   |   a   |  aa   |   0     |   0   |   1   |   0   |
+// |  3  |   b   |   b   |  bb   |   0     |   0   |   1   |   0   |
+// |  4  |   c   |   c   |  cc   |   0     |   0   |   1   |   0   |
+// |  5  |   aa  |   bb  |  cc   |   0     |   1   |   0   |   0   |
+// |-----|-------|-------|-------|---------|-------|-------|-------|
+//
+// Any advice value that appears in multiple rows has the consistency of its value enforced across
+// rows via permutation argument, e.g. row #0 `a` == row #2 `a` is enforced within in the
+// permutation argument.
+
+use halo2::{
+    circuit::{layouter::SingleChipLayouter, Cell, Chip, Layouter, Region},
+    dev::MockProver,
+    pasta::Fp,
+    plonk::{
+        Advice, Any, Assignment, Circuit, Column, ConstraintSystem, Error, Permutation, Selector,
+    },
+    poly::Rotation,
+};
+
+// A value that has been allocated in the constraint system.
+struct Alloc {
+    cell: Cell,
+    value: Fp,
```

Comment:
As mentioned, this variable needs to support parameter generation when values are unknown:
```suggestion
    value: Option<Fp>,
```